File: app/scraper/sync.py
import logging
from datetime import UTC, datetime
from typing import Any

from app.models import Listing, PriceHistory

logger: logging.Logger = logging.getLogger(name=__name__)


async def upsert_listings(listings: list[Listing]) -> tuple[list[Listing], list[Listing]]:
    """Upsert listings into the database.

    Returns (new_listings, price_changed_listings).
    """
    new_listings: list[Listing] = []
    price_changed: list[Listing] = []

    for listing in listings:
        existing = await Listing.find_one(Listing.yad2_id == listing.yad2_id)

        if existing is None:
            # New listing
            await listing.insert()
            new_listings.append(listing)

            # Record initial price
            if listing.price:
                await PriceHistory(
                    listing_id=listing.yad2_id,
                    price=listing.price,
                ).insert()
        else:
            # Update existing listing
            existing.last_seen_at = datetime.now(tz=UTC)
            existing.is_active = True

            # Check for price change
            if listing.price and existing.price and listing.price != existing.price:
                old_price = existing.price
                existing.price = listing.price
                existing.price_per_sqm = listing.price_per_sqm

                await PriceHistory(
                    listing_id=existing.yad2_id,
                    price=listing.price,
                ).insert()

                if listing.price < old_price:
                    price_changed.append(existing)

                logger.info(msg=f'Price change for {existing.yad2_id}: {old_price} -> {listing.price}')

            # Update other fields that might have changed
            existing.description = listing.description or existing.description
            existing.images = listing.images or existing.images
            if listing.amenities:
                existing.amenities = listing.amenities
            if listing.location:
                existing.location = listing.location

            await existing.save()

    logger.info(msg=f'Upsert complete: {len(new_listings)} new, {len(price_changed)} price drops')
    return new_listings, price_changed


async def deduplicate_listing(listing: Listing) -> bool:
    """Check if a listing is a duplicate of an existing one (republished with new ID).

    Returns True if it's a duplicate.
    """
    if not listing.address.street or not listing.rooms:
        return False

    # Look for existing listing with same address + rooms + sqm but different yad2_id
    query: dict[str, Any] = {
        'yad2_id': {'$ne': listing.yad2_id},
        'address.city': listing.address.city,
        'address.street': listing.address.street,
        'rooms': listing.rooms,
        'is_active': True,
    }

    if listing.sqm:
        query['sqm'] = listing.sqm

    existing = await Listing.find_one(query)
    return existing is not None

File: next/src/lib/scraper/sync.ts
import { createServerClient } from "@/lib/supabase/server";
import type { ListingInsert } from "@/types";
import type { Marker } from "@/lib/scraper/yad2-client";
import { parseMarker } from "@/lib/scraper/yad2-client";
import type { DealType } from "@/types";

export interface UpsertResult {
  newListings: ListingInsert[];
  priceDrops: Array<{ listing: ListingInsert; oldPrice: number }>;
}

/**
 * Upsert a batch of parsed listings into Supabase.
 * - New listings: insert + record initial price history
 * - Existing listings: update last_seen_at, detect price changes
 */
export async function upsertListings(listings: ListingInsert[]): Promise<UpsertResult> {
  const supabase = createServerClient();
  const newListings: ListingInsert[] = [];
  const priceDrops: Array<{ listing: ListingInsert; oldPrice: number }> = [];

  if (listings.length === 0) return { newListings, priceDrops };

  // Fetch existing listings by yad2_id for price comparison
  const yad2Ids = listings.map((l) => l.yad2_id);
  const { data: existing } = await supabase
    .from("listings")
    .select("yad2_id, price")
    .in("yad2_id", yad2Ids);

  const existingMap = new Map<string, number | null>();
  for (const row of existing ?? []) {
    existingMap.set(row.yad2_id, row.price);
  }

  const toInsert: ListingInsert[] = [];
  const toUpdate: Array<{ yad2_id: string; data: Partial<ListingInsert> }> = [];
  const priceHistoryInserts: Array<{ listing_id: string; price: number }> = [];

  for (const listing of listings) {
    const existingPrice = existingMap.get(listing.yad2_id);

    if (existingPrice === undefined) {
      // New listing
      toInsert.push(listing);
      newListings.push(listing);

      if (listing.price) {
        priceHistoryInserts.push({
          listing_id: listing.yad2_id,
          price: listing.price,
        });
      }
    } else {
      // Existing listing — update
      const updateData: Partial<ListingInsert> = {
        last_seen_at: new Date().toISOString(),
        is_active: true,
      };

      // Price change detection
      if (listing.price && existingPrice && listing.price !== existingPrice) {
        updateData.price = listing.price;
        updateData.price_per_sqm = listing.price_per_sqm;

        priceHistoryInserts.push({
          listing_id: listing.yad2_id,
          price: listing.price,
        });

        if (listing.price < existingPrice) {
          priceDrops.push({ listing, oldPrice: existingPrice });
        }
      }

      // Update fields that may have changed
      if (listing.description) updateData.description = listing.description;
      if (listing.images && listing.images.length > 0) updateData.images = listing.images;
      if (listing.location) updateData.location = listing.location;

      toUpdate.push({ yad2_id: listing.yad2_id, data: updateData });
    }
  }

  // Batch insert new listings (upsert to handle race conditions)
  if (toInsert.length > 0) {
    const { error } = await supabase
      .from("listings")
      .upsert(toInsert, { onConflict: "yad2_id" });
    if (error) {
      console.error("Error inserting listings:", error.message);
    }
  }

  // Batch update existing listings
  for (const { yad2_id, data } of toUpdate) {
    await supabase.from("listings").update(data).eq("yad2_id", yad2_id);
  }

  // Insert price history records
  if (priceHistoryInserts.length > 0) {
    const { error } = await supabase.from("price_history").insert(priceHistoryInserts);
    if (error) {
      console.error("Error inserting price history:", error.message);
    }
  }

  console.log(
    `Upsert complete: ${newListings.length} new, ${priceDrops.length} price drops`
  );
  return { newListings, priceDrops };
}

/**
 * Process raw markers into listings and upsert them.
 */
export async function processMarkerChunk(
  markers: Marker[],
  dealType: DealType
): Promise<UpsertResult> {
  const listings: ListingInsert[] = [];
  for (const marker of markers) {
    const listing = parseMarker(marker, dealType);
    if (listing) listings.push(listing);
  }
  return upsertListings(listings);
}

/**
 * Check if a listing is a duplicate (same address + rooms + sqm, different yad2_id).
 */
export async function deduplicateListing(listing: ListingInsert): Promise<boolean> {
  if (!listing.street || !listing.rooms) return false;

  const supabase = createServerClient();
  let query = supabase
    .from("listings")
    .select("yad2_id")
    .neq("yad2_id", listing.yad2_id)
    .eq("city", listing.city ?? "")
    .eq("street", listing.street ?? "")
    .eq("rooms", listing.rooms)
    .eq("is_active", true)
    .limit(1);

  if (listing.sqm) {
    query = query.eq("sqm", listing.sqm);
  }

  const { data } = await query;
  return (data?.length ?? 0) > 0;
}
